torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -529,10 +529,10 @@ void TensorPipeAgent::shutdownImpl() {
   if (timeoutThread_.joinable()) {
     timeoutThread_.join();
   }
-  // TODO: context_->join() is not absolutely ready yet.
-  // NOTE: context_->join() will wait for available RPC message to be
-  //       read or written, and wait for the remaining unavailable ones
-  //       to be called with error by invoking callbacks.
+
+  // This will close all the pipes and listeners, invoke all callbacks with
+  // errors, turn down the I/O event loops and wait for everything to terminate.
+  context_->join();
```

Comment:
Now I understand your reasoning, and I think either is OK. It's up to you whether it should be in the current state or revert to the previous version.